tensorflow/compiler/xla/service/pattern_matcher.h

```diff
@@ -1266,6 +1266,59 @@ class HloInstructionPatternOpcodeImpl {
   bool invert_;
 };
 
+// An HloInstructionPattern implementation that matches if the instruction has
+// an opcode from a prescribed list or is identical to the operand.
+template <typename OperandType, typename OperandImpl>
+class HloInstructionPatternOpcodeOrOperandImpl {
+ public:
+  explicit HloInstructionPatternOpcodeOrOperandImpl(
+      absl::Span<const HloOpcode> opcodes,
+      const HloInstructionPattern<OperandType, OperandImpl>& operand,
+      bool invert)
+      : opcodes_(opcodes), operand_(operand), invert_(invert) {}
+
+  bool Match(::xla::HloInstruction* inst, MatchOption option) const {
+    // Transparent case matches the operand.
+    if (operand_.Match(inst, option)) {
+      return true;
+    }
+    // Attempt to match entries of opcodes_.
+    bool match = c_linear_search(opcodes_, inst->opcode());
+    if (invert_ && match) {
+      EXPLAIN << " HloInstruction also has one of opcodes {"
+              << HloOpcodeString(opcodes_) << "}, expected anything else.";
+      return false;
+    }
+    if (!invert_ && !match) {
+      EXPLAIN << " HloInstruction also doesn't have any of opcodes {"
+              << HloOpcodeString(opcodes_) << "}.";
+      return false;
+    }
+    // Operands must match.
```

Comment:
This comment isn't clear to me.  The two operands in question here are not the same sorts of thing -- one is a pattern, and the other is an operand of the unary expression.  So in this sentence we're comparing apples and oranges, so to speak.